0000-plugins.md

```diff
@@ -0,0 +1,162 @@
+	...
+  }
+
+//export create_local_post
+func my_plugin() int32 {
+	params := CreatePost{}
+	// Read post data
+	err := pdk.InputJSON(&params)
+	if err != nil {
+		pdk.SetError(err)
+		return 1
+	}
+	if strings.Contains(params.Body, "Java") {
+		// Throw error to reject post
+		pdk.SetError("We dont talk about Java)
+		return 1
+	}
+	params.Body = strings.Replace(params.Body, "Rust", "Go", -1);
+	// Write back post data
+	err = pdk.OutputJSON(params)
+	if err != nil {
+		pdk.SetError(err)
+		return 1
+	}
+	return 0
+}
+
+func main() {}
+```
+
+## Host Functions
+
+Plugins will need access to the Lemmy database for various reasons, such as determining which instance a new post is coming from, or storing custom data. The easiest way to do this would be direct database access for plugins, but this has many problems:
+- Database schema changes across Lemmy versions, which would make it very difficult to make plugins compatible with new Lemmy versions, and prevent instances from upgrading
+- Plugins could accidentally delete data or break the database
+
+As such we need to use a different approach, by exposting specific database calls from Lemmy via [host functions](https://extism.org/docs/concepts/host-functions). For the beginning the following functions will be exposed:
+
+- Instance::read_or_create 
+
+More functions may be added if they are needed for the initial plugins.
+
+## Initial Plugins
+
+These can mainly serve as proof of concept, and as examples which can be modified by plugin developers. Include integration tests to ensure that functionality works with new Lemmy changes. These plugins should be written in a variety of languages, and live in a separate repo `lemmy-plugin-examples`.
+
+- Block only downvotes from a specific instance
+- Replace words in post or comment text (eg "cloud to butt")
+- Replace Youtube URLs in post links with [Invidious](https://invidious.io/)
+- Push notifications (send data of new posts to localhost url, [#2631](https://github.com/LemmyNet/lemmy/issues/2631))
+- Language detection (automatically apply language tag to posts/comments, [#2870](https://github.com/LemmyNet/lemmy/issues/2870))
+- Archive previous versions of posts when they are edited
+- Reject comments from specific user IDs (like [!santabot@slrpnk.net](https://slrpnk.net/post/11069853))
+- Additional suggestions welcome
+
+
+## Plugin Licensing
+
+Plugins can use any license chosen by the developer, including proprietary. This ensures that plugins can integrate with proprietary libaries, for example Android/iOS push notifications.
+
+# Drawbacks
+
+Plugins may have a negative impact on Lemmy's performance. It is up to plugin developers and instance admins to ensure that plugins run adequately. In any case plugins are only called for POST actions but not for GET. This means slow plugins might result in longer waiting time to submit posts, but passive browsing will be unaffected.
+
+# Rationale and alternatives
+
+TODO
+
+# Prior art
+
```

Comment:
Plugins to federated servers:

* https://github.com/friendica/friendica-addons
* https://codeberg.org/hubzilla/hubzilla-addons
* https://nextcloud.github.io/news/features/plugins/
* https://developer.wordpress.org/plugins/

Plugins to federated clients:

* hundreds of frameworks for all previous real time chat systems

Reply:
Thanks Ive added those links.